torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -284,6 +296,12 @@ def _pre_forward(
     # the `grad_fn` is mutated.
     _register_post_backward_hooks(state, handles)
 
+    # Recursively convert args and kwargs to specified precision.
+    input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
```

Comment:
I see that this line only targets the wrapper FSDP code path.

I recommend adding a `TODO` at this line to indicate that we need to change it to support the non-wrapper code path; however, we may also file an issue if you do not want to re-run CI.

---
The open problem is how should `fully_shard` enable applying different `fsdp_kwargs` to different `FlatParamHandle` (for a select subset of the arguments). The main difference is that `fully_shard` does not support manual wrapping, which is the solution for `FullyShardedDataParallel`. Today, one application of `fully_shard` does not interact with any other application of `fully_shard`. If we want to migrate to the composable APIs, then we need to iron out the design here.

Given a away to specify different `fsdp_kwargs` to different `FlatParamHandle`, then the change to this line here is not complicated.

For the non-wrapper code path, we change the requirement to be that every handle in `handles` (passed to this `_pre_forward`) has the same low precision parameter dtype. In today's design (and probably for the near future), `handles` is a singleton (i.e. `len(handles) == 1`). Either way, the general constraint will be on `handles`.

```
def _get_handles_param_dtype(handles: Iterable[FlatParamHandle]) -> Optional[torch.dtype]:
    """
    Assumes all ``handles`` have the same low precision parameter dtype
    setting, where ``None`` represents full precision. This returns the single
    low precision parameter dtype setting across ``handles``.
    """
    dtypes: Optional[torch.dtype] = set(handle._config.low_prec_param_dtype for handle in handles)
    p_assert(
        len(dtypes) == 1,
        f"Expects uniform low precision parameter dtype but got {dtypes}",
    )
    return next(iter(dtypes))        
```
```
input_dtype: Optional[torch.dtype] = _get_handles_param_dtype(handles)
```

(The question is how to specify `handle._config._low_prec_param_dtype` to be different across `FlatParamHandle`s of the same `fully_shard`.)

Reply:
> Today, one application of fully_shard does not interact with any other application of fully_shard. If we want to migrate to the composable APIs, then we need to iron out the design here.

Can we do the following:

For short term, we require users to wrap children modules with `fully_shard` before wrapping parent ones. When applying `fully_shard`, recursively check all submodules' states and skip the ones that already have `replicate`, `fully_shard` on it. More specifically, `fully_shard` can write a flag in every submodule, indicating that this submodule is handled by a parent module already. If users later applied another `fully_shard` on the submodule, error out. 

For the longer term, when we have the high-level wrapper API, that API can sort things out, and we no longer require users to wrap things in order.